File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    arr = [[0] * i for i in range(1, n+1)]
    directions = [[1, 0], [0, 1], [-1, -1]] # 아래, 오른쪽, 왼쪽위(좌측 대각선 위)
    turn = 0
    x, y = 0, 0
    i = 1
    end_num = sum(i for i in range(1, n+1))
    
    while i <= end_num:
        arr[x][y] = i
        i += 1
        mx, my = x + directions[turn][0], y + directions[turn][1]
        
        if 0 <= mx < n and 0 <= my < n and arr[mx][my] == 0:
            x, y = mx, my
        else:
            turn = (turn + 1) % 3
            x += directions[turn][0]
            y += directions[turn][1]
     
    answer = []
    for row in arr:
        for item in row:
            answer.append(item)
    return answer
    

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
  const arr = Array.from({ length: n }, (_, i) => Array(i + 1).fill(0));
  const directions = [
    [1, 0],   // 아래
    [0, 1],   // 오른쪽
    [-1, -1], // 왼쪽위 (좌측 대각선 위)
  ];

  let turn = 0;
  let x = 0, y = 0;
  let i = 1;
  const endNum = (n * (n + 1)) / 2;

  while (i <= endNum) {
    arr[x][y] = i++;
    const [dx, dy] = directions[turn];
    const mx = x + dx;
    const my = y + dy;

    if (mx >= 0 && mx < n && my >= 0 && my <= mx && arr[mx][my] === 0) {
      x = mx;
      y = my;
    } else {
      turn = (turn + 1) % 3;
      x += directions[turn][0];
      y += directions[turn][1];
    }
  }

  const answer = [];
  for (const row of arr) {
    for (const num of row) {
      answer.push(num);
    }
  }

  return answer;
}
